### PR TITLE
Ignite 4188

### PR DESCRIPTION
**Pull request by SomeFire**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicCache.java [hunk 1]
```diff
@@ -2849,7 +2849,7 @@ else if (readers.contains(nearNode.id())) // Reader became primary or backup.
 
             while (true) {
                 for (int i = 0; i < req.size(); i++) {
-                    GridDhtCacheEntry entry = entryExx(req.key(i), topVer);
+                    GridDhtCacheEntry entry = entryExx(req.key(i), topVer, 0);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicCache.java]: We have to make method with default parameter.
So we can avoid unnecessary changes

```
public XXX entryExx(Key key, TopVer topVer) {
  return entryExx(key, topVer, 0);
}
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 2]
```diff
@@ -149,6 +150,20 @@ public GridDhtColocatedCache(GridCacheContext<K, V> ctx, GridCacheConcurrentMap
...
+
+        ctx.io().addHandler(ctx.cacheId(), GridNearUnlockResponse.class, new CI2<UUID, GridNearUnlockResponse>() {
+            @Override public void apply(UUID nodeId, GridNearUnlockResponse res) {
+                processUnlockResponse(nodeId, res);
+            }
+        });
+    }
+
+    private void processUnlockResponse(UUID id, GridNearUnlockResponse res) {
+        SavepointUnlockFutureGridDhtColocatedLockFuture2 fut = (SavepointUnlockFutureGridDhtColocatedLockFuture2) ctx.mvcc().future(res.futureId());
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: SavepointUnlockFutureGridDhtColocatedLockFuture2 - please, make class name more readable.
What it for?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 3]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
...
+                KeyCacheObject cacheKey = key instanceof IgniteTxKey ? ((IgniteTxKey) key).key() : ctx.toCacheKeyObject(key);
+                IgniteTxKey txKey = key instanceof IgniteTxKey ? (IgniteTxKey) key : ctx.txKey(cacheKey);
+
+                GridDistributedCacheEntry entry = peekExx(cacheKey);
+
+                GridCacheMvccCandidate lock =
+                    ctx.mvcc().removeExplicitLock(Thread.currentThread().getId(), txKey, null);
+
+                if (lock != null) {
+
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: empty line not necessary here @

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 4]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
...
+                    final AffinityTopologyVersion topVer = lock.topologyVersion();
+
+                    assert topVer.compareTo(AffinityTopologyVersion.ZERO) > 0;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Can `topVer` be null?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 5]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
         }
     }
 
+    /** {@inheritDoc} */
+    @Override public void unlockAllForSavepoint(@Nullable Collection<? extends K> keys, GridCacheVersion cacheVer,
+        GridNearTxLocal tx) throws IgniteCheckedException {
+        if (keys.isEmpty())
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: `keys` is nullable but there is no null check.
You can use `F.isEmpty` for null safe emptiness check

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 6]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
...
+                        if (log.isDebugEnabled())
+                            log.debug("Failed to unlock keys (all partition nodes left the grid).");
+
+                        continue;
+                    }
+
+                    if (map == null) {
+                        Collection<ClusterNode> affNodes = CU.allNodes(ctx, topVer);
+
+                        keyCnt = (int)Math.ceil((double)keys.size() / affNodes.size());
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Why map initialization inside loop?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 7]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
...
+                    if (!lock.reentry()) {
+                        if (!ver.equals(lock.version()))
+                            throw new IgniteCheckedException("Failed to unlock (if keys were locked separately, " +
+                                "then they need to be unlocked separately): " + keys);
+
+                        if (!primary.isLocal()) {
+                            GridNearUnlockRequest req = map.get(primary);
+
+                            if (req == null) {
+                                SavepointUnlockFutureGridDhtColocatedLockFuture2 fut = new SavepointUnlockFutureGridDhtColocatedLockFuture2(
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: May be we need simple constructor with ctx and tx instead of mess with all params?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java [hunk 8]
```diff
@@ -802,6 +818,210 @@ else if (log.isDebugEnabled())
...
+                    ClusterNode primary = ctx.affinity().primaryByKey(key, topVer);
+
+                    if (primary == null) {
+                        if (log.isDebugEnabled())
+                            log.debug("Failed to unlock keys (all partition nodes left the grid).");
+
+                        continue;
+                    }
+
+                    if (map == null) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Code duplication. Map initialization should be outside loop.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 2 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Renamed to SavepointUnlockFuture. It is used to wait until all locks for rolled back keys will be unlocked.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 5 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 3 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 4 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Only if lock is broken. I'm not sure, but I think lock branch is never accessible during savepoint unlocks. Need more tests to check it.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 6 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Cleared and moved outside loop.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 8 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Cleared and moved outside loop.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java (hunk 7 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedCache.java]: Optimized a bit. But we need some of them (like timout).

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicCache.java (hunk 1 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/atomic/GridDhtAtomicCache.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java [hunk 9]
```diff
@@ -462,7 +462,7 @@ public boolean dumpDebugInfo() {
      * @return Cached affinity.
      */
     public AffinityAssignment cachedAffinity(AffinityTopologyVersion topVer) {
-        if (topVer.equals(AffinityTopologyVersion.NONE))
+        if (topVer != null && topVer.equals(AffinityTopologyVersion.NONE))
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java]: Seems that change doesn't belong to current issue.
Need to revert.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+                    tx.rollbackToSavepoint("sp");
+
+                    IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            return cache.putIfAbsentAsync(1, 1).get();
+                        }
+                    },cacheMode().name()+"_get");
+
+                    assertTrue(fut.get());
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java]: fut.get - have to specify timeout for future.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TransactionProxyImpl.java [hunk 11]
```diff
@@ -421,4 +421,57 @@ private void saveFuture(IgniteInternalFuture<IgniteInternalTx> fut) {
     @Override public String toString() {
         return S.toString(TransactionProxyImpl.class, this);
     }
+
+    /** {@inheritDoc} */
+    @Override public void savepoint(String name) throws IllegalArgumentException {
+        savepoint(name, false);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TransactionProxyImpl.java]: Please explain, why `overwrite` flag is false here?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 12]
```diff
@@ -148,6 +150,9 @@
     /** */
     protected CacheWriteSynchronizationMode syncMode;
 
+    /** List of savepoints for this transaction, which is used to retrieve previous states. */
+    protected LinkedList<TxSavepointLocal> savepoints;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Why `LinkedList` here?
Can we use `HashMap<String, TxSavepointLocal>` here?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 13]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
...
+                " Savepoint will not be created.");
+
+            return;
+        }
+
+        ListIterator<TxSavepointLocal> spIter = findSP(name);
+
+        if (spIter != null) {
+            if (overwrite) {
+                releaseSP(spIter);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Why we have several methods for one operation?
`releaseSP`, `releaseSPs`, `releaseSavepoint`.
Seems code need to be refactored to use only one method to release savepoint.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 14]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
...
+            } else {
+                throw new IllegalArgumentException("Savepoint \"" + name + "\" already exist. " +
+                    "Savepoints with the same name aren't available. " +
+                    "If you want to rewrite savepoints - use savepoint(name, true) method.");
+            }
+        }
+
+        addSP(new TxSavepointLocal(name, this));
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Method `addSP` used only from here. Can be inlined.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 15]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
...
+
+        if (log.isDebugEnabled())
+            log.debug("Saving point created [savepoint=" + name + ", tx=" + this + ']');
+
+        incSavepointOrder();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Method can be inlined.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxLocal.java [hunk 16]
```diff
@@ -632,4 +633,12 @@ public void rollbackDhtLocal() throws IgniteCheckedException {
     @Override public String toString() {
         return GridToStringBuilder.toString(GridDhtTxLocal.class, this, "super", super.toString());
     }
+
+    void clearTxState(GridCacheVersion ver) {
+        ((IgniteTxStateImpl)txState).clearDhtEntries(ver.savepointOrder());
+
+        ((IgniteTxStateImpl)txState).refreshReadWriteViews();
+
+        updateSavepointOrder(ver.savepointOrder() + 1);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxLocal.java]: Method used only here. Can be inlined.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 17]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
...
         protected abstract IgniteInternalFuture<T> postMiss(T t) throws IgniteCheckedException;
     }
+
+    /**
+     * Creates savepoint.
+     *
+     * @param name Savepoint ID.
+     */
+    @Override public void savepoint(String name, boolean overwrite) {
+        if (currentPrepareFuture() != null) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Here and in other places we have to use `state()` method to check transaction `state() == ACTIVE`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 18]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
...
+            log.info("Trying to create savepoint after commit process was started." +
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Is it `info` log level? Don't we need some exception?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java [hunk 19]
```diff
@@ -1772,4 +1777,149 @@ protected PostLockClosure1(T arg, boolean commit) {
          */
         protected abstract IgniteInternalFuture<T> postMiss(T t) throws IgniteCheckedException;
     }
+
+    /**
+     * Creates savepoint.
+     *
+     * @param name Savepoint ID.
+     */
+    @Override public void savepoint(String name, boolean overwrite) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: On start of every user operation with transaction have to call `checkValid()` method.
To ensure timeout, state and other checks

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java (hunk 9 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java]: Yeah, it was made during unlockAllForSavepoint creating and I forgot to revert it when working method was finished.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java (hunk 10 above)
Comment (SomeFire) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TransactionProxyImpl.java (hunk 11 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TransactionProxyImpl.java]: We use 'false' as default parameter to keep same behaviour as databases mentioned in the ticket.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 12 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: We need ordered collection to move forward/backward through savepoints. LinkedhashMap may be? But is it reasonable?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 19 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 17 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 18 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: I don't think we need exception when transaction already finished by commit or rollback. Savepoint methods just will do nothing.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 14 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: I'm not sure. `savepoint` method only call other methods and do nothing by itself. Inlining `addSP` will mess different layers: method calls and direct actions.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java (hunk 15 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxLocalAdapter.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxLocal.java (hunk 16 above)
Comment (SomeFire) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxLocal.java]: I'm not finished with savepoints, I think I will use this method somewhere else.
